seat.c

```diff
@@ -273,21 +270,170 @@ handle_keyboard_key(struct wl_listener *listener, void *data)
 	struct cg_seat *seat = keyboard->seat;
 
-	wl_list_remove(&keyboard->destroy.link);
-	wl_list_remove(&keyboard->modifiers.link);
-	wl_list_remove(&keyboard->key.link);
 	wl_list_remove(&keyboard->link);
-	free(keyboard);
-
+	wl_list_remove(&keyboard->destroy.link);
+	cg_keyboard_destroy(keyboard);
 	update_capabilities(seat);
 }
 
+static void
+handle_keyboard_group_key(struct wl_listener *listener, void *data) {
+	struct cg_keyboard_group *cg_group =
+	    wl_container_of(listener, cg_group, key);
+	handle_key_event(cg_group->keyboard, data);
+}
+
+static void
+handle_keyboard_group_modifiers(struct wl_listener *listener, void *data) {
+	struct cg_keyboard_group *group =
+	    wl_container_of(listener, group, modifiers);
+	handle_modifier_event(group->keyboard);
+}
+
+static bool
+keymaps_match(struct xkb_keymap *km1, struct xkb_keymap *km2) {
+	char *km1_str = xkb_keymap_get_as_string(km1, XKB_KEYMAP_FORMAT_TEXT_V1);
+	char *km2_str = xkb_keymap_get_as_string(km2, XKB_KEYMAP_FORMAT_TEXT_V1);
+	bool result = strcmp(km1_str, km2_str) == 0;
+	free(km1_str);
+	free(km2_str);
+	return result;
+}
+
+static bool
+repeat_info_match(struct cg_keyboard *a, struct wlr_keyboard *b) {
+	return a->device->keyboard->repeat_info.rate == b->repeat_info.rate &&
+	       a->device->keyboard->repeat_info.delay == b->repeat_info.delay;
+}
+
+struct cg_keyboard *
+cg_keyboard_from_seat(struct cg_seat *seat, struct wlr_input_device *device) {
+	struct cg_keyboard *keyboard = calloc(1, sizeof(struct cg_keyboard));
+	if (keyboard == NULL) {
+		wlr_log(WLR_ERROR, "Could not allocate new cg_keyboard.");
+		return NULL;
+	}
+
+	keyboard->seat = seat;
+	keyboard->device = device;
+
+	return keyboard;
+}
+
+static void
+keyboard_group_add(struct cg_keyboard *keyboard) {
+	struct cg_seat *seat = keyboard->seat;
+	struct wlr_keyboard *wlr_keyboard = keyboard->device->keyboard;
+	struct cg_keyboard_group *group;
+	wl_list_for_each(group, &seat->keyboard_groups, link) {
+		struct wlr_keyboard_group *wlr_group = group->wlr_group;
+		if (keymaps_match(wlr_keyboard->keymap, wlr_group->keyboard.keymap) &&
+		   repeat_info_match(keyboard, &wlr_group->keyboard)) {
+			wlr_log(WLR_DEBUG, "Adding keyboard %p to group %p.", (void*)keyboard,
+			        (void*)wlr_group);
+			wlr_keyboard_group_add_keyboard(wlr_group, wlr_keyboard);
+			return;
+		}
+	}
+	/* This is reached if and only if the keyboard could not be inserted into
+	 * any group */
+	struct cg_keyboard_group *cg_group =
+	    calloc(1, sizeof(struct cg_keyboard_group));
+	if (cg_group == NULL) {
+		wlr_log(WLR_ERROR, "Failed to allocate keyboard group.");
+		return;
+	}
+	cg_group->wlr_group = wlr_keyboard_group_create();
+	if (cg_group->wlr_group == NULL) {
+		wlr_log(WLR_ERROR, "Failed to create wlr keyboard group.");
+		goto cleanup;
+	}
+
+	cg_group->wlr_group->data = cg_group;
+	wlr_keyboard_set_keymap(&cg_group->wlr_group->keyboard,
+	                        keyboard->device->keyboard->keymap);
```

Comment:
Sway is also setting the repeat info here, why isn't this necessary in Cage?

Reply:
No, your right, this is necessary. I forgot it since cage keyboards don't have a repeat_info, but instead I just used the `wlr_keyboard` values now.